src/compiler/checker.ts

```diff
@@ -36120,7 +36120,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 }
                 return getUnaryResultType(operandType);
             case SyntaxKind.ExclamationToken:
-                checkTruthinessExpression(node.operand);
+                checkTruthinessOfType(operandType, node.operand);

```

Comment:
This direct call to `checkTruthinessOfType` when the operand type is known mirrors our handling of binary expressions in `onOperator`.

I checked all other call sites, and only this one benefits from not reevaluating the type.